user groups: Prevent user not in user group to edit/delete group name/members.

Kumar-Tarun opened the pull request:
This PR fixes #8315. It prevents users who are not a part of user group to edit/delete user group name/members. Realm admins and user group members can continue to do the above operations. 
I'm attaching two gifs showing the behavior of user group UI after changes, also if a user creates a group, adds some members in it, then removes himself/herself, he/she cannot further edit/delete the user group, the code prevents doing that.
The following picture has Cordelia as current user.
![user groups](https://user-images.githubusercontent.com/23361315/36146710-88e5acd2-10db-11e8-8bb6-bfc29577a9c3.gif)
The following picture has iago as current user (admin).
![1](https://user-images.githubusercontent.com/23361315/36146709-87dcf17e-10db-11e8-99ba-dcdd0a9bc14f.gif)

Kumar-Tarun commented:
@hackerkid Can you review the behavior now regarding user groups ? I've made settings such that, even if one changes html/css temporarily through dev tools of browser, one cannot change contents of the group, and cannot definitely update backend.

hackerkid commented:
@Kumar-Tarun You have to enforce the check in the backend as well.

Kumar-Tarun commented:
@hackerkid Okay, I'll do that and update the PR, btw is the behavior fine?

Kumar-Tarun commented:
@hackerkid I'll do that definitely. I need some time, since I have some tests scheduled in my college, but I will do it. To enforce tests in backend, I thought of enforcing in the user_groups.py in zerver/views, is that fine, or should I do it some where else (like in the user_groups.py in zerver/lib) ?

hackerkid commented:
@Kumar-Tarun I thinks views is good enough.

Kumar-Tarun commented:
@hackerkid Please take a look. I've added backend enforcing and tests as well. Sorry for the delay.

Kumar-Tarun commented:
@hackerkid I agree that would be better, so should I remove individual settings and add them for the whole box?

hackerkid commented:
@Kumar-Tarun Yup. Make sure to backup the code in case we want to go back :)

Kumar-Tarun commented:
@hackerkid I've updated the PR, but unfortunately I could replace only one line of code, the other not-allowed cursors had to be kept because the user-group box property was not applying to individual elements, nevertheless it now works in the desired way! :)

hackerkid commented:
@Kumar-Tarun The cursor not allowed icon is still not enabled in all of the box.
![p8365](https://user-images.githubusercontent.com/7190633/36689493-9763d99e-1b55-11e8-9433-b37a302af42e.gif)

Kumar-Tarun commented:
@hackerkid wait let me see it.

hackerkid commented:
@Kumar-Tarun Hold on. I forgot to fetch your branch. Its working fine now :)

hackerkid commented on zerver/lib/user_groups.py:
```diff
@@ -74,6 +74,12 @@ def create_user_group(name: Text, members: List[UserProfile], realm: Realm,
 
+def get_members_user_group(user_group: UserGroup) -> List[UserProfile]:
+    members = UserGroupMembership.objects.filter(
+        user_group=user_group,
```
This can be 1 line instead of 3 lines right?

hackerkid commented on zerver/lib/user_groups.py:
```diff
@@ -74,6 +74,12 @@ def create_user_group(name: Text, members: List[UserProfile], realm: Realm,
+    members = UserGroupMembership.objects.filter(
+        user_group=user_group,
+    )
+    return list(member.user_profile.id for member in members)
```
Isn't `return [member.user_profile.id for member in members]` simpler?

hackerkid commented on zerver/views/user_groups.py:
```diff
@@ -50,6 +54,13 @@ def edit_user_group(request: HttpRequest, user_profile: UserProfile,
+    user_group = access_user_group_by_id(user_group_id, realm=user_profile.realm)
+    group_member_ids = get_members_user_group(user_group)
+
+    if (user_profile.id not in group_member_ids and not user_profile.is_realm_admin):
```
Check for user_profile.is_realm_admin first as it is (1) and the other is (N)

hackerkid commented on zerver/views/user_groups.py:
```diff
@@ -93,5 +108,14 @@ def remove_members_from_group_backend(request: HttpRequest, user_profile: UserPr
+            raise JsonableError(_("There is no member %s in this user group" % (member,)))
+
+    if (user_profile.id not in group_member_ids and not user_profile.is_realm_admin):
+        raise JsonableError(_("Only user group members and admins are allowed to remove members from it"))
```
No need of `from it`

hackerkid replied:
admins -> organization administrators.

hackerkid commented on zerver/views/user_groups.py:
```diff
@@ -77,8 +88,12 @@ def add_members_to_group_backend(request: HttpRequest, user_profile: UserProfile
     existing_member_ids = set(get_memberships_of_users(user_group, user_profiles))
+
+    if (user_profile.id not in group_member_ids and not user_profile.is_realm_admin):
+        raise JsonableError(_("Only user group members and admins are allowed to add members to groups"))
```
No need of `to groups`

hackerkid replied:
admins -> organization administrators.

hackerkid commented on zerver/views/user_groups.py:
```diff
@@ -35,6 +35,10 @@ def edit_user_group(request: HttpRequest, user_profile: UserProfile,
     user_group = access_user_group_by_id(user_group_id, realm=user_profile.realm)
+    group_member_ids = get_members_user_group(user_group)
+
+    if (user_profile.id not in group_member_ids and not user_profile.is_realm_admin):
```
Check for user_profile.is_realm_admin first as it is (1) and the other is (N)

Fix other occurances as well

hackerkid commented on zerver/tests/test_user_groups.py:
```diff
@@ -140,6 +139,17 @@ def test_user_group_update(self) -> None:
         result = self.client_patch('/json/user_groups/1111', info=params)
         self.assert_json_error(result, "Invalid user group")
 
+        self.logout()
```
Add tests here as well as in other cases to make sure that administrators can modify the groups

hackerkid commented on static/templates/user-groups-admin.handlebars:
```diff
@@ -1,5 +1,8 @@
 <div id="user-groups-admin" class="settings-section" data-name="user-groups-admin">
     <div class="tip">Anyone in this organization can add user groups.</div>
+    {{#unless is_admin}}
+    <div class="tip">Only user group members can modify their respective group name/description/members.</div>
```
Is't this better?
`Only user group members and organization administrators can modify the group.`

hackerkid replied:
I think showing `Only organization administrators can edit these settings` here doesn't make any sense. What do you think?
![screenshot from 2018-02-27 01-47-53](https://user-images.githubusercontent.com/7190633/36693128-4c3b10ee-1b60-11e8-9aa7-946e0c1773d9.png)

hackerkid replied:
There is no setting for disabling users from adding new groups right?. In that case we can delete this box as well.
`Anyone in this organization can add user groups.` Have a separate commit for this alert box changes including the one I suggested before.

Kumar-Tarun replied:
@hackerkid thanks for this detailed review, I will surely update these as soon as I can.
I agree with you on the last two points regarding the removal of boxes, they don't make sense in the current scenario.

Kumar-Tarun commented:
@hackerkid I've updated the PR, please have a look! :)

showell commented on static/js/settings_user_groups.js:
```diff
@@ -234,7 +265,7 @@ exports.set_up = function () {
         var name = $(this).siblings('.name').text().trim();
         var btn = $(this);
-
+        if (user_groups.is_member_of(group_id, people.my_current_user_id()) || me.is_admin) {
```
The indentation here is off.

Kumar-Tarun replied:
@showell I've updated the PR, you may check!

Kumar-Tarun commented:
@showell @hackerkid can you please give this a look ?, I've updated it and resolved the merge conflicts.

showell commented:
@Kumar-Tarun I think you can do more on the JS side to make the diffs smaller while keeping them atomic.  I gave some examples on czo of how you can wrap pieces of code in functions that you execute immediately to reduce variable scope without changing the behavior of the code.  These can be good prep commits to get merged quickly.

showell commented:
@Kumar-Tarun Is there anything you can do to break up the first commit into smaller commits?  You should be able to split out commits where you introduce the function wrappers without changing anything else.

showell commented on frontend_tests/node_tests/settings_user_groups.js:
```diff
@@ -268,6 +289,163 @@ set_global('people', {
+            $('.user-group').attr('id', '1');
+            assert.equal(fn.call(fake_delete), undefined);
+        }
+        if (event_name === 'keypress' && child_selector === '.user-group h4 > span') {
```
You don't really need the `keypress` section here.  I tried removing it and the tests still pass.  (You don't need to return a `noop` function, because nothing is going to execute this anyway.)

Another thing to note is that `zjquery` actually has a default `.on(...)` function that you can probably just let run.  And then when you want to test the individual handlers you can do something like this:

~~~
settings_user_group.set_up();
delete_handler = $('#user_groups').get_on_handler('.click', '.delete.');
delete_handler();

Kumar-Tarun replied:
@showell since settings_user_groups.set_up() is being called twice in whole testing, due to which 'on' is being set twice due to which get_on_handler is working only for first time and is interfering in tests of test_on_events(), so I think it's better to have custom 'on' method in my case, to prevent conflicting tests.
Regarding keypress part, I'll remove that.

showell replied:
You should just reset all of the simulated jQuery before doing `set_up()` the second time.  This way you'll prevent all possible leaks.  If you need to duplicate some setup, just make functions for those.

~~~
set_global('$', global.make_zjquery());

Kumar-Tarun replied:
@showell  I found a shorter method, instead of resetting zjquery, since I just need to remove duplicate 'on' functions, isn't it better that I keep the 'on' click handlers when set_up() was first called, and when I call it the second time, I still use get_on_handlers, but just simulate the 'on' to return nothing, in turn preventing it register duplicate 'on' handlers. I think it faster and easy to implement and will reduce writing duplicate code. Let me know what you think!

showell replied:
I would just reset zjquery every time you call `set_up`.  Resetting zjquery isn't expensive at all, and leaks are a pain to debug.

showell replied:
I understand your objection here, but I already addressed it two comments ago.  If there is setup that you need to repeat both times you do setup, you can de-duplicate this by creating functions.You can create a function named something like `reset_test_setup` that does the following:

* reset zjquery
* call `settings_user_groups.set_up`
* build up some fake DOM elements

Basically, all the common stuff can be put in a function.  All the non-common stuff needs to be non-common, anyway.

Kumar-Tarun replied:
So, then call this function, every time before doing set_up() ?

showell replied:
Thanks!  The function wrappers make the subsequent diff a lot more readable!

showell commented on frontend_tests/node_tests/settings_user_groups.js:
```diff
@@ -268,6 +289,163 @@ set_global('people', {
+            var fake_change = $.create('fk-#user-groups_click_save');
+            fake_change.set_parents_result('.user-group', $('.user-group'));
+            $('.user-group').attr('id', '3');
+            assert.equal(fn.call(fake_change), undefined);
```
It seems like you should be making more assertions here about what the function actually does.

Kumar-Tarun replied:
@showell here I can't make any other assertions since main functionality is already being tested in test_on_events function, also since I am just returning from that click handler, so I can only assert that return value is undefined.

showell replied:
If you're not making any meaningful assertions, then I would just remove the test code, since it's gonna give a false sense of assurance in the coverage reports.  Or make helper functions that you can call in both paths.

rishig commented:
Quick visual design notes:
* Let's remove the trash can in cases where the user cannot edit
* Ditto for the "Add members .." text
* Possibly let's also remove the box around the pills, in this case?
* We don't need the :no_entry_sign: for the mouse

Thanks!

Kumar-Tarun commented:
@rishig see which one you like better - 
![14](https://user-images.githubusercontent.com/23361315/37309069-c5f1538e-2665-11e8-9583-b66965059cb6.PNG)


![13](https://user-images.githubusercontent.com/23361315/37309072-c91a3404-2665-11e8-96e8-1e74ade49c6d.PNG)

rishig commented:
This is for the "users cannot edit" case? I like the second.

rishig commented:
By the way, are the pills here different from the ones used in composing a PM?

Kumar-Tarun commented:
@rishig Ok so I will go with the second one, and yes pills are a little different from PM pills, also, Vishnu and I think that 'not allowed cursor' will give a enhanced feel as to saying that user is not allowed to modify, but if you say, I'll remove that as well. Let me know!

Kumar-Tarun commented:
@showell I've updated the PR, you may check, also I implemented the suggestions as said by Rishi, except one of them, 'not allowed cursor' which I wanted to just confirm, because I think keeping it would be a little verbose in hinting user that he/she is not allowed to modify group.

rishig commented:
I think as the final state we won't want the "not allowed cursor", since we don't use it anywhere else in the product, and we also want to explain why the user is not able to edit if there is a possibility of confusion. After playing around with it I agree with your intuition that there is a possibility of confusion, so we'll want to do something.

For this PR, let's leave it with the "not allowed cursor", and then do a followup project to
* sort the groups the user can edit to the top
* have a tip, that says something like "Please contact an administrator or a member of the group to be added to any of the groups below." (will have to figure out exact wording) right above where the non-editable section starts.

rishig commented:
oh I see, we use the not-allowed-cursor on the custom emoji page, which this page is modeled after. I think the use there is also not great, but we can fix that afterwards.

Kumar-Tarun commented:
@rishig that's a great idea to be worked on after this project, and will definitely improve user experience, also, we also use not allowed cursor in organization profile page, from which I got the idea, nevertheless thanks for this idea! :)

timabbott commented:
@Kumar-Tarun you'll need to do some updates to this following #8716.

timabbott commented on static/js/settings_user_groups.js:
```diff
@@ -76,40 +75,43 @@ exports.populate_user_groups = function () {
+                },
+                stopAdvance: true,
+            });
+        }());
```
I'm not sure this improves the codebase; this whole block is already wrapped in a function, and none of these define new variables.

Kumar-Tarun replied:
https://github.com/zulip/zulip/pull/8365#issuecomment-372025849, Steve and I thought this, and also we thought that it would be good to some early exit conditions, also see this -https://chat.zulip.org/#narrow/stream/91-code-review/subject/PR.20.238365 :)
sorry for the delay, I'm out of city right now.

timabbott commented on static/templates/user-groups-admin.handlebars:
```diff
@@ -1,5 +1,8 @@
     <div class="tip">Anyone in this organization can add user groups.</div>
+    {{#unless is_admin}}
+    <div class="tip">Only user group members and organization administrators can modify the group.</div>
+    {{/unless}}
```
Should this not show if the user is a member?

Kumar-Tarun replied:
Here, this warning needs to be shown for all users who are not admins, since it might be that user is member of one group and not of other, and its beneficial to warn them of the rules.

rishig replied:
If we show this, minor edits:
"Only group members and organization administrators can modify a group."

Though I think once we sort them into "Your Groups" and "Other groups", we won't need this warning at all.

Kumar-Tarun replied:
That's right, for the time being, let's display the above warning, when we will sort them, we'll remove that then.

Kumar-Tarun commented:
@timabbott I've updated the PR, and also left some comments, you may check and let me know, if anything else is required. :)

timabbott commented on static/styles/input_pill.css:
```diff
@@ -32,6 +32,20 @@
+.notmem .pill:focus {
+    color: hsl(177, 38%, 54%);
+    border: 1px solid #b2e4e2;
+}
```
These should be scoped inside `.pill-container` -- we don't like having CSS that by appearance could affect other parts of the site.

timabbott commented:
I merged the first commit and posted comments on the second.

timabbott commented on zerver/views/user_groups.py:
```diff
@@ -35,6 +35,10 @@ def edit_user_group(request: HttpRequest, user_profile: UserProfile,
+    group_member_ids = get_user_group_members(user_group)
+    msg = _("Only user group members and organization administrators are allowed to edit groups")
+    if (not user_profile.is_realm_admin and user_profile.id not in group_member_ids):
+        return json_error(msg)
```
This code block is duplicated in several places.  You should put it into a function, and have it raise `JsonableError` (an exception which results in an equivalent `json_error`) 

Or, better, we should refactor `access_user_group_by_id` to accept a user object rather than a realm

timabbott commented:
This works great BTW.  We should hide the "x" styling in the `.notmem` input-pill case though; it's confusing.

Also you should re-order the commits; the two tip commits should be squashed together, and the backend commit should come before the tip commit.

@synicalsyntax can you give the JS part of this a review?  I'd love it if we had a way to remove the redundant `exports.can_edit(data.id, me)` calls.

Kumar-Tarun commented:
@timabbott I've updated the PR, and made the changes you suggested.

Kumar-Tarun commented on static/js/settings_user_groups.js:
```diff
@@ -61,6 +66,28 @@ exports.populate_user_groups = function () {
         });
 
+        function update_membership(group_id) {
+            if (!exports.can_edit(group_id, me)) {
```
Good idea, didn't think that! Thanks!

Kumar-Tarun commented on static/js/settings_user_groups.js:
```diff
@@ -179,67 +204,90 @@ exports.populate_user_groups = function () {
         }
 
         $('#user-groups #' + data.id).on('blur', '.input', function (event) {
+            if (!exports.can_edit(data.id, me)) {
```
@synicalsyntax thanks for this one too! I'll update the PR accordingly.

timabbott commented on zerver/views/user_groups.py:
```diff
@@ -26,6 +26,11 @@ def add_user_group(request: HttpRequest, user_profile: UserProfile,
+def error_msg(user_profile: UserProfile, user_group: UserGroup, msg: Text) -> Any:
+    group_member_ids = get_user_group_members(user_group)
+    if (not user_profile.is_realm_admin and user_profile.id not in group_member_ids):
+        return JsonableError(msg)
```
This should just raise the JsonableError rather than returning; them you don't need to have the callers do `if error is not None`.

I'd also move this into `access_user_group_by_id` (and passing in a `UserProfile`) rather than having a separate 
function.

Finally, I don't think we need separate error messages for each case.  It's a waste of translators' time.  We should just have all 4 of these cases be

"Only group members and organization administrators can administer this user group."

rishig replied:
possible amendment: I think it might be reasonable to use "group" instead of "user group" when the context is clear. So "Only group members and organization administrators can administer this group."?

Kumar-Tarun replied:
@rishig I'll do that.

Kumar-Tarun commented:
@timabbott I've updated the PR, you may check!

timabbott commented:
I've merged this, thanks @Kumar-Tarun and @synicalsyntax!  

That said, I think refactoring is in order to clean up the JS code.  I think this function in particular has no hope of being maintained correctly over time:

```
 
+        function update_membership(group_id) {
+            if (exports.can_edit(group_id)) {
+                return;
+            }
+            userg.find('.name').attr('contenteditable','false');
+            userg.find('.description').attr('contenteditable','false');
+            userg.addClass('ntm');
+            pill_container.find('.input').attr('contenteditable','false');
+            pill_container.find('.input').css('display', 'none');
+            pill_container.addClass('notmem');
+            pill_container.off('keydown', '.pill');
+            pill_container.off('keydown', '.input');
+            pill_container.off('click');
+            pill_container.on('click', function (e) {
+                e.stopPropagation();
+            });
+            pill_container.find('.pill').hover(function () {
+                pill_container.find('.pill').find('.exit').css('opacity', '0.5');
+                    }, function () {}
+            );
+        }
```

So I'm opening a follow-up issue to clean up this section of the codebase.  I think we might want to extract out a "non-editable" pill concept, and just have a user group's section be rerendered completely (replacing the old content) when the current user joins/leaves a group; done right, we shouldn't need a function like this at all.
